### PR TITLE
ENH: move an exception and add a prehook to check for exception place…

### LINKED ISSUE #27656: API: many custom errors / warnings are not exposed in pandas.errors
From https://github.com/pandas-dev/pandas/issues/27553 and https://github.com/pandas-dev/pandas/issues/27522, I quickly looked at how consistent we are with defining errors and warnings in `pandas.errors`.

So we have a `pandas.errors` module which is meant to publicly expose our custom exceptions and warnings.

Some are defined there such as PerformanceWarning, UnsortedIndexError, ParserError, ... (https://github.com/pandas-dev/pandas/blob/bb6135880e5e453d7701764b9f2e4ad3356a68d7/pandas/errors/__init__.py). 
But many are not:

* tslibs.parsing.DateParseError
* tslibs.period.IncompatibleFrequency
* pandas/_config/config.py: OptionError
* pandas/core/base.py: DataError, GroupByError, SpeciicationError
* pandas/core/common.py: SettingWithCopyError/Warning
* pandas/core/computation/common/py: NameResolutionError
* pandas/core/computation/engine.py: NumExprClobberingError
* pandas/core/computation/ops.py: UndefinedVariableError
* pandas/core/indexes/base.py: InvalidIndexError
* pandas/coreindexing.py: IndexingError
* pandas/io/clipboard/exceptions.py: PyperclipException
* pandas/io/formats/css.py: CSSWarning
* pandas/io/msgpack: several exceptions
* pandas/io/pytables.py: lots of exceptions
* pandas/io/sql.py: SQLALchemyRequired, DatabaseError
* pandas/io/stata.py: several warnings

Do we want to move them all to the public `pandas.errors` module? 
Are there reasons that some can / should not be moved? Eg should the cython ones be defined in the cython file for performance? 
Are there certain custom exceptions/warnings that we rather convert in a built-in one instead of publicly exposing them? 
Do we want to move all the io related ones? (those modules are somewhat public)

### PR DESCRIPTION
**Pull request by dataxerik**
- [x] closes #27656.
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

This is a follow up from my last MR: https://github.com/pandas-dev/pandas/pull/47901. I'm new to prehooks, so I may have went in the wrong direction for the prehook. But, I was thinking of checking the files for class that inherit Exceptions and flagging them. Let me know if another approach is preferred.

I also noticed some other exceptions while working on this.

- InvalidVersion. I tried moving this one, but got a circular import so I'm importing it in the errors.__init__.py instead.
- LossySetitemError, NoBufferPresent, and InvalidComparison are moved to errors.__init__.py.
- NotThisMethod. It seem like a code generated exception rather than a custom pandas one, so I'm importing it in the errors.__init__.py instead.

### CODE REVIEW

File: scripts/validate_exception_location.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+class Visitor(ast.NodeVisitor):
+    def __init__(self, path: str) -> None:
+        self.path = path
+
+    def visit_ClassDef(self, node):
+        classes = {getattr(n, "id", None) for n in node.bases}
+
+        if (
+            classes
+            and classes.issubset(exception_warning_list)
```
Comment (mroeschke) [inline scripts/validate_exception_location.py:92]: Not great we have to keep another list of pandas defined warnings. Can we check if the defined Warning/Exception came from `pandas/`?

Comment (dataxerik) [reply scripts/validate_exception_location.py:92]: Sorry for sleeping on this for so long. I'm not sure if I understand the question.

It's good if we can remove `permisable_exception_warning_list`, correct? I don't like it either, but I not sure how to handle preexisting exception outside of `errors/__init__.py` and `_libs`. For example, I could tell `LossySetitemError` came from `pandas/core/dtypes/cast.py`. But, if someone added another exception to `cast.py` then I'm not sure how to flag the new exception, but allow the old exception. 

Does that get at the question or am I off base?

Comment (mroeschke) [reply scripts/validate_exception_location.py:92]: > It's good if we can remove permisable_exception_warning_list, correct?

Yeah this is my main concern. IMO any pandas error (whether just internal or exposed externally) should be imported into `pandas/errors/__init__.py`

Maybe one approach would be

1. Grab all the warnings/exceptions from `doc/source/reference/testing.rst` (this file should be in sync with `errors/__init__.py` through another check
2. AST parse through `pandas/` (excluding tests) to record all defined warning/exception subclasses
3. Compare the results.

Comment (dataxerik) [reply scripts/validate_exception_location.py:101]: Gotcha, thank you for the help and the clarification. I've updated the MR to leverage `testing.rst`. Let me know if this solution works better

File: doc/source/whatsnew/v1.5.0.rst [hunk 2]
```diff
@@ -317,7 +317,7 @@ Other enhancements
...
+- :class:`.DataError`, :class:`.SpecificationError`, :class:`.SettingWithCopyError`, :class:`.SettingWithCopyWarning`, :class:`.NumExprClobberingError`, :class:`.UndefinedVariableError`, :class:`.IndexingError`, :class:`.PyperclipException`, :class:`.PyperclipWindowsException`, :class:`.CSSWarning`, :class:`.PossibleDataLossError`, :class:`.ClosedFileError`, :class:`.IncompatibilityWarning`, :class:`.AttributeConflictWarning`, :class:`.DatabaseError`, :class:`.PossiblePrecisionLoss`, :class:`.ValueLabelTypeMismatch`, :class:`.InvalidColumnName`, :class:`.CategoricalConversionWarning`, :class:`.InvalidComparison`, :class:`.InvalidVersion`, :class:`.LossySetitemError`, :class:`.NoBufferPresent`, and :class:`.NotThisMethod` are now exposed in ``pandas.errors`` (:issue:`27656`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:320]: Could you move the newly added ones to `v.1.6.0.rst`. Sorry, the 1.5.0.rc must have been released while this was worked on

File: pandas/tests/test_errors.py [hunk 3]
```diff
@@ -11,33 +11,38 @@
 @pytest.mark.parametrize(
     "exc",
     [
-        "UnsupportedFunctionCall",
-        "UnsortedIndexError",
-        "OutOfBoundsDatetime",
-        "ParserError",
-        "PerformanceWarning",
+        "AttributeConflictWarning",
```
Comment (mroeschke) [inline pandas/tests/test_errors.py:14]: Guessing this is just alphabetized correct?

File: scripts/tests/test_validate_exception_location.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+def my_func():
+  pass
+
+class {custom_name}({error_type}):
+  pass
+
+"""
+
+# Test with various python-defined exceptions to ensure they are all flagged.
+testdata = [
```
Comment (mroeschke) [inline scripts/tests/test_validate_exception_location.py:26]: Could make this a pytest fixture that is parameterized over these values instead.

File: scripts/validate_exception_location.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
+"""
+Validate that the exceptions and warnings are in approrirate places.
```
Comment (mroeschke) [inline scripts/validate_exception_location.py:2]:
```suggestion
Validate that the exceptions and warnings are in appropriate places.
```

File: scripts/validate_exception_location.py (hunk 1 above)
Comment (mroeschke) [reply scripts/validate_exception_location.py:101]: This line can be removed now right? The Warning/Exception name should be just checked against `self.exception_set` now?

Comment (dataxerik) [reply scripts/validate_exception_location.py]: I think I still need that line. I sort of kept my original approach. This approach was to have `Visitor` class check if the class was a warning/exception subclass and check if the class was in the testing.rst. Now, I've reworked to be more inline with your approach. So, `Visitor` gets all the subclass exceptions and then `validate_exception_and_warning_placement` will use `self.exception_set` to find the difference. 

Let me know if that is reasonable/make sense

File: doc/source/whatsnew/v1.5.0.rst (hunk 2 above)
Comment (dataxerik) [reply doc/source/whatsnew/v1.5.0.rst]: Sorry about that. I've updated this

File: pandas/tests/test_errors.py (hunk 3 above)
Comment (dataxerik) [reply pandas/tests/test_errors.py:14]: I meant to call that out in the PR. But, yeah, I just alphabetized them

File: scripts/tests/test_validate_exception_location.py (hunk 4 above)
Comment (dataxerik) [reply scripts/tests/test_validate_exception_location.py]: I think I changed this right. Thank you for the suggestion. I didn't know I could it that way

File: scripts/validate_exception_location.py (hunk 5 above)
Comment (dataxerik) [reply scripts/validate_exception_location.py]: Updated

File: pandas/errors/__init__.py [hunk 6]
```diff
@@ -12,6 +12,9 @@
     OutOfBoundsTimedelta,
 )
 
+from pandas._version import NotThisMethod
```
Comment (mroeschke) [inline pandas/errors/__init__.py:15]: Just realized this is from our vendored version of versioneer which the user should never see. In the pre-commit check could you ignore `versioneer.py`?

File: scripts/validate_exception_location.py (hunk 1 above)
Comment (dataxerik) [reply scripts/validate_exception_location.py]: Yeah, I don't really like the giant list either. AFAIK, the base is a `ast.Name` object, and that really only has an `id` which acts as the name: https://docs.python.org/3/library/ast.html#ast.Name. 

However, if we're comfortable assuming that if a base name is `Exception` or if the name ends with either `Error` or `Warning` then that's an `Exception` or  `Warning`.  I think we could re-write it like the below. It's slightly less hard-codey. Let me know if that shows good/reasonable. 

```
    def visit_ClassDef(self, node) -> None:
        def is_an_exception_subclass(base_id: str) -> bool:
            return base_id == "Exception" or base_id.endswith("Warning") or base_id.endswith("Error")

        classes = []
        for n in node.bases:
            base_id = getattr(n, "id", None)
            if base_id and is_an_exception_subclass(base_id):
                classes.append(base_id)

        if classes:
            self.possible_exceptions.add(node.name)
```

File: pandas/errors/__init__.py (hunk 6 above)
Comment (dataxerik) [reply pandas/errors/__init__.py:15]: I believe that's coming from `pandas/_version.py`, but yeah I can ignore `pandas/_version.py`

File: scripts/validate_exception_location.py (hunk 1 above)
Comment (mroeschke) [reply scripts/validate_exception_location.py]: Hmm I tried something like this and was able to get the name of the base class

```
In [1]: import ast

In [2]: text = """
   ...: class MyWarning(Warning):
   ...:     pass
   ...:
   ...: class MyClass:
   ...:     pass
   ...: """

In [12]: class Visit(ast.NodeVisitor):
    ...:     def visit_ClassDef(self, node):
    ...:         breakpoint()
    ...:         return node
    ...:

In [13]: p = ast.parse(text)

In [14]: Visit().visit(p)
> <ipython-input-12-85c0ebda4faf>(4)visit_ClassDef()
-> return node
(Pdb) node
<ast.ClassDef object at 0x125468760>
(Pdb) node.bases
[<ast.Name object at 0x125468640>]
(Pdb) node.bases[0].id
'Warning'

```

Comment (dataxerik) [reply scripts/validate_exception_location.py]: I could be off-base, but sounds like we are on the same page. I'm grabbing the ids from the bases and comparing the values. I've update the MR as well. Let me know if I'm missing something

Comment (mroeschke) [reply scripts/validate_exception_location.py]: Ah gotcha. Yeah from the newer diff this is what I had in mind. Thanks!

**Comment by mroeschke** [comment]
Looked pretty good. Could you fix the merge conflict and see if the Ci failures are related?

**Comment by dataxerik** [comment]
> 



> Looked pretty good. Could you fix the merge conflict and see if the Ci failures are related?

Sounds good. I've merged main

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Awesome work on this @dataxerik! This will really help make sure users are able to publicly access Warnings and Exceptions in the future!

**Comment by jorisvandenbossche** [comment]
(just getting here from the notification that the issue was closed, so sorry for the late comment)

I _think_ that `InvalidComparison` is only used internally to raise certain cases which is then catched internally again to do something else or raise a public error. Assuming that's the case, we don't need to expose this exception class publicly? 

(the same might be true for `LossySetitemError`, didn't check in detail, cc @jbrockmendel)

**Comment by mroeschke** [comment]
That's a fair point. 

Maybe we can clearly document the exceptions or warnings that are supposed to be internal? My concern is that overtime the "internal" exceptions or warnings may unintentionally become publicly thrown, and at least users could still catch them through `pandas.errors`

**Comment by jbrockmendel** [comment]
> I think that InvalidComparison is only used internally to raise certain cases which is then catched internally again to do something else or raise a public error. Assuming that's the case, we don't need to expose this exception class publicly?

correct.  It should always be caught and re-raised.  Could clarify that in the docstring.

Ditto for LossySetitemError.

**Pull request merged**